providers/azure.go

```diff
@@ -230,44 +276,48 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 		err = otherMailsErr
 	}
 
+	if err != nil || email == "" {
+		email, err = json.Get("userPrincipalName").String()
+		if err != nil {
+			logger.Errorf("unable to find userPrincipalName: %s", err)
+			return "", err
+		}
+	}
+
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	var email string
-	var err error
-
-	if s.AccessToken == "" {
+func getEmailFromProfileAPI(ctx context.Context, accessToken, profileURL string) (string, error) {
+	if accessToken == "" {
 		return "", errors.New("missing access token")
 	}
 
-	json, err := requests.New(p.ProfileURL.String()).
+	json, err := requests.New(profileURL).
 		WithContext(ctx).
-		WithHeaders(makeAzureHeader(s.AccessToken)).
+		WithHeaders(makeAzureHeader(accessToken)).
 		Do().
 		UnmarshalJSON()
 	if err != nil {
 		return "", err
 	}
 
-	email, err = getEmailFromJSON(json)
-	if err == nil && email != "" {
-		return email, err
-	}
+	return getEmailFromJSON(json)
+}
 
-	email, err = json.Get("userPrincipalName").String()
-	if err != nil {
-		logger.Errorf("failed making request %s", err)
-		return "", err
+// EnrichSessionState finds the email to enrich the session state
+func (p *AzureProvider) EnrichSessionState(ctx context.Context, s *sessions.SessionState) error {
```

Comment:
The interface method is named `EnrichSession` now -- I promise I won't refactor it anymore 😅 

Can we move this up to the top? I find it helpful for people consuming providers to have the methods in the same order that the `OAuthProxy` will call them in for the login flow: (GetLoginURL, Redeem, EnrichSession). I'm trying to shift to that organization as I refactor more providers

Reply:
sure! renamed and move them up